### PR TITLE
Fix `GPSampler`'s suggestion failure within `torch.no_grad()` context manager

### LINKED ISSUE #5663: GPSampler's suggestion fails when used inside `torch.no_grad()` context manager
### Expected behavior

Currently, GPSampler's `sample_relative` fails when used inside `torch.no_grad()` context manager, resulting the default initial kernel params to be used.
Specifically, the `torch.no_grad()` context manager, when used outside of Optuna, affects the PyTorch gradient calculations within GPSampler and raises a error, causing `sample_relative` to be fail.

The expected behavior is:

- GPSampler work correctly even when  the `torch.no_grad()` is active.

### Environment

- Optuna version: 4.0.0
- Python version: 3.11.0
- OS: MacOS, Ubuntu
- torch: 2.4.0

### Error messages, stack traces, or logs

```shell
[I 2024-09-03 07:44:07,293] A new study created in memory with name: no-name-2ddc8765-755f-4bca-8a7a-fd0f3249e467
[I 2024-09-03 07:44:07,322] Trial 0 finished with value: 0.32613587379455566 and parameters: {'n_hidden': 3}. Best is trial 0 with value: 0.32613587379455566.
[W 2024-09-03 07:44:07,560] The optimization of kernel_params failed: element 0 of tensors does not require grad and does not have a grad_fn
The default initial kernel params will be used instead.
[I 2024-09-03 07:44:07,801] Trial 1 finished with value: -0.08969029039144516 and parameters: {'n_hidden': 10}. Best is trial 1 with value: -0.08969029039144516.
```

### Steps to reproduce

The following code yields the error above:

```python
import optuna
import torch
from torch import nn

class Net(nn.Module):
   def __init__(self, n_input, n_output, n_hidden):
       super(Net, self).__init__()
       self.fc1 = nn.Linear(n_input, n_hidden)
       self.fc2 = nn.Linear(n_hidden, n_output)

   def forward(self, x):
       x = torch.relu(self.fc1(x))
       x = self.fc2(x)
       return x

def objective(trial: optuna.Trial) -> float:
   with torch.no_grad():
       n_input = 10
       n_output = 1
       n_hidden = trial.suggest_int('n_hidden', 1, 10)
       net = Net(n_input, n_output, n_hidden)
       x = torch.randn(1, n_input)
       y = net(x)
   return y.item()

sampler = optuna.samplers.GPSampler(n_startup_trials=1)
study = optuna.create_study(sampler=sampler)
study.optimize(objective, n_trials=2)
```

### Additional context (optional)

_No response_

### PR DESCRIPTION
**Pull request by kAIto47802**
## Motivation

This PR aims to:

- Resolve #5663 .

## Description of the changes

- Set grad enabled with in the `loss_func` in `_fit_kernel_params` function in `optuna/_gp/gp.py`.
- Add unit test for `_fit_kernel_params` function.

### CODE REVIEW

**Comment by eukaryo** [comment]
@contramundum53 @not522 Could you review this PR?

**Review by not522 — COMMENTED** [review]
Thank you for your PR.

Your test seems a bit complex for unit testing and can be split into some tests, e.g., the bug with `no_grad` is fixed, the kernel params are correct, etc. This PR should focus on fixing the `no_grad` bug, so checking that the returned `kernel_params` differs from the `initial_kernel_params` may be enough.

And please follow the naming rule for test files, i.e., `gp_tests/test_gp.py` is the right place to implement it.

**Comment by kAIto47802** [comment]
Thank you for your review.
I have fixed them. Specifically, I have:

- removed the check whether the kernel params are correct, and
- renamed test files to `tests/gp_tests/test_gp.py`.

**Review by contramundum53 — APPROVED** [review]
LGTM!

**Review by not522 — APPROVED** [review]
LGTM
